west: Add `west rtt` command

topisani opened the pull request:
Implemented in the pyocd and openocd runners while awaiting some feedback.

This command runs separately from a debug session. We should probably also support attaching a rtt client to a running debugserver. However, this might work differently for each runner. pyocd is easy, as it supports a separate RTT command, while openocd and jlink will always need a running server.

See the FIXME's for other points i would like input on.

topisani committed:
west: runners: Add `west rtt` command with pyocd implementation

This command runs separately from a debug server, instead of attaching
to a running server. This is both the easiest out of the box experience,
and also should be possible to implement consistently for most runners.

This commit includes an initial implementation for pyocd.

Signed-off-by: Tobias Pisani <mail@topisani.dev>

MarkoSagadin commented:
This would be really awesome to have. For our internal processes I created a [wrapper](https://github.com/IRNAS/irnas-east-software/blob/main/src/east/system_commands/jlink_commands.py) around JLinkExeRTTViewer so we could have a single command that does that.

How would enabling a running server be implemented in the case of JLink? As a separate command or an additional flag to the `west flash` command?

topisani commented:
> How would enabling a running server be implemented in the case of JLink? As a separate command or an additional flag to the `west flash` command?

I think the two possible approaches, of which we should maybe do both as separate commands, are:
 1. `west rtt -r openocd` will run the `openocd` server just like `west debugserver` does, and attach an rtt client to it.
 2. `west rtt -r openocd` requires you to run `west debugserver` in another terminal and connects to that.
 
 The first approach is easy, when all you want is to see the log, but the second is more flexible and compatible with the other debug/debugserver/attach commands.

MarkoSagadin commented:
I think that second approach makes sense, like you said, it is compatible with other commands.

Two things:

1. Is explicit `-r openocd` flag needed always? Or just in this PR? I would expect that I can just run `west rtt` regardless of the runner, if the default one support this, of course.
2. What about if the `west rtt` command checks if there is a running server (by checking known ports maybe?) and creates a new one (like suggested in your first case) if none is found? This would then support the easy use case that you describe.  So something like:
    ```
    west build -b <some_board> && west flash && west rtt
    ```

topisani commented:
>     1. Is explicit `-r openocd` flag needed always? Or just in this PR? I would expect that I can just run `west rtt` regardless of the runner, if the default one support this, of course.

No, i just always use it, because the default runner is usually not what i want
 
>     2. What about if the `west rtt` command checks if there is a running server (by checking known ports maybe?) and creates a new one (like suggested in your first case) if none is found? This would then support the easy use case that you describe.  So something like:

I consideren this option, but i think its nicer to have it either as a separate command, or as an argument. Connecting to arbitrary ports when not asked to is probably a no go. Besides, consistent behavior seems preferable

topisani committed:
west: runners: Implement the west rtt command for openocd

This was non-trivial, as openocd is a bit weird to work with. Using only
commands passed with '-c' arguments, I couldn't get it to reliably resume
(or just not halt) the target when started. I tried using the 'sleep'
command, and various 'configure -event XX { resume }' events, but nothing
panned out, as it seems to always halt after all `-c` commands have been
run.

To avoid that, this waits for the TCL RPC port to be up, and sends a
resume command there. This works reliably.

Signed-off-by: Tobias Pisani <mail@topisani.dev>

pdgendt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -12,6 +12,11 @@
 from os import path
 from pathlib import Path
 from zephyr_ext_common import ZEPHYR_BASE
+import socket
+import selectors
+import sys
+import time
+import shutil
```
Nit: maybe put the import in alphabetical order as you're adding a bunch of new ones.

pdgendt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
+            raise ValueError("RTT Control block not be found")
+
+        rtt_cmds = [
+            '-c', 'rtt setup 0x{:x} 0x10 "SEGGER RTT"'.format(rtt_address),
+            '-c', 'rtt server start {} 0'.format(self.rtt_port),
```
Use formatted strings

```suggestion
            '-c', f'rtt setup 0x{rtt_address:x} 0x10 "SEGGER RTT"',
            '-c', f'rtt server start {self.rtt_port} 0',
```

pdgendt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
+        ]
+
+        server_cmd = (self.openocd_cmd + self.cfg_cmd +
+                      ['-c', 'tcl_port {}'.format(self.tcl_port),
+                       '-c', 'telnet_port {}'.format(self.telnet_port),
+                       '-c', 'gdb_port {}'.format(self.gdb_port)] +
```
Same here

```suggestion
                      ['-c', f'tcl_port {self.tcl_port}',
                       '-c', f'telnet_port {self.telnet_port}',
                       '-c', f'gdb_port {self.gdb_port}'] +
```

pdgendt commented on scripts/west_commands/runners/pyocd.py:
```diff
@@ -213,3 +215,23 @@ def debug_debugserver(self, command, **kwargs):
+        if rtt_addr is None:
+            raise ValueError('RTT control block not found')
+
+        self.logger.debug('rtt address: 0x{:x}'.format(rtt_addr))
+
+        cmd = ([self.pyocd] +
+               ['rtt'] +
+               self.pyocd_config_args +
+               self.daparg_args +
+               self.target_args +
+               self.board_args +
+               self.frequency_args +
+               self.tool_opt_args +
+               ['-a', '0x{:x}'.format(rtt_addr)])
```
Same here

pdgendt commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
+
+        if self.thread_info_enabled and self.supports_thread_info():
+            pre_init_cmd.append("-c")
+            rtos_command = '${} configure -rtos Zephyr'.format(self.target_handle)
```
And here

pdgendt commented:
Looks very neat BTW

topisani committed:
west: runners: Implement the west rtt command for jlink

Moves the telnet client into runners/core.py as well, as this is now shared
between openocd and jlink.

Signed-off-by: Tobias Pisani <mail@topisani.dev>

topisani commented:
LMK if i should have pushed this latest commit to a different PR, now that this had been partially approved. I just thought id better get this jlink support in along with the rest.

jlink support is implemented in much the same way as openocd, except it can better automatically locate the rtt block, so it relies on that.

fabiobaltieri reviewed (approved):
Cool, second commit message is stale I guess, feel free to update it.

topisani commented:
> Cool, second commit message is stale I guess, feel free to update it.

Any objections to me just squashing all three commits?

fabiobaltieri commented:
> Any objections to me just squashing all three commits?

Yes, don't. :-) Just reword the second commit message, the rest is fine as it is.

topisani commented:
alright, reworded and reordered, apparently i had swapped the first two commits somewhere along the way

pdgendt commented:
A follow-up PR to add this to the [documentation](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#west-runner) would be nice.
